### PR TITLE
auditor: initial commit

### PR DESCRIPTION
**Pull request by slint**
* Adds the ZenodoAuditor module, responsible for auditing record,
  file and served content consistency.

Signed-off-by: Alexander Ioannidis <a.ioannidis@cern.ch>

### CODE REVIEW

File: zenodo/modules/auditor/utils.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    """Return a ZenodoRecord generator with all records."""
+    records = (db.session.query(RecordMetadata)
+               .join(PersistentIdentifier,
+                     RecordMetadata.id == PersistentIdentifier.object_uuid)
+               .filter(PersistentIdentifier.pid_type == 'recid',
+                       PersistentIdentifier.status == PIDStatus.REGISTERED))
+    return (ZenodoRecord(data=r.json, model=r) for r in records)
+
+
+class tree(dict):
```
Comment (krzysztof) [inline zenodo/modules/auditor/utils.py]: you can also use:
```
from collections import defaultdict
def tree():
    return defaultdict(tree)
```

Comment (slint) [inline zenodo/modules/auditor/utils.py]: Originally I used and preferred this for its brevity, but from [this SO answer](http://stackoverflow.com/a/19829714) I changed my mind. The `defaultdict` implementation had issues with serialization, comparisons, etc.

File: zenodo/modules/auditor/oai.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    def __init__(self, audit_id, logger, communities):
+        """Initialize an OAI audit."""
+        super(OAIAudit, self).__init__(audit_id, logger)
+
+        self._build_oai_sets_from_db()
+        self._checks = chain(iter((OAICorrespondenceCheck(),)),
+                             (OAISetResultCheck(self, c) for c in communities))
+
+    def _build_oai_sets_from_db(self):
+        self.redis = StrictRedis()
```
Comment (slint) [inline zenodo/modules/auditor/oai.py]: :+1: I didn't notice that it was a generic Flask-Cache. I'm using Redis sets though, so I'll have to assert Redis for the `CACHE_TYPE`.

File: zenodo/modules/auditor/records.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        # Maybe use 'checksum', 'file_id', 'version_id'
+        files = self.record.get('_files', [])
+        duplicate_files = duplicates([f['key'] for f in files])
+        if duplicate_files:
+            self.issues['files']['duplicates'] = duplicate_files
+
+    def _multiple_buckets(self):
+        files = self.record.get('_files', [])
+        buckets = {f['bucket'] for f in files}
+        if len(buckets) != 1:
```
Comment (krzysztof) [inline zenodo/modules/auditor/records.py]: ``>`` + check for missing _files

File: zenodo/modules/auditor/ext.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    def __init__(self, app=None):
+        """Extension initialization."""
+        if app:
+            self.init_app(app)
+
+    def init_app(self, app):
+        """Flask application initialization."""
+        from .cli import audit
+        app.cli.add_command(audit)
```
Comment (krzysztof) [inline zenodo/modules/auditor/ext.py]: add in ``zenodo/cli.py`` instead https://github.com/zenodo/zenodo/blob/master/zenodo/cli.py#L36-L37

File: zenodo/modules/auditor/oai.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+          - Elasticsearch Records Index
+          - The `/oai2d` Endpoint
+
+        Checking that the set of identifiers is consistent throughout these
+        sources is vital.
+        """
+        db_ids = self._db_identifiers()
+        es_ids = self._es_identifiers()
+        oai2d_ids = self._oai2d_endpoint_identifiers()
+
```
Comment (krzysztof) [inline zenodo/modules/auditor/oai.py]: S =(es_id | oai2d_ids | db_ids)
for name, X in zip(('oai','es','db', (oai, es, db)):
  self.issues[name] = S - X

**Pull request merged**